File: carma/util/NotificationConsumer.cc

// $Id: NotificationConsumer.cc,v 1.42 2013/01/03 21:19:53 abeard Exp $

/**
 * A base class to accept notifications from a Push supplier. This class
 * is usually extended and the push method overridden.
 *
 *
 *  @author Chul Gwon
 *  @version $Revision: 1.42 $, $Date: 2013/01/03 21:19:53 $
 */

#include "carma/util/NotificationConsumer.h"

#include "carma/util/UserException.h"
#include "carma/util/programExtras.h"
#include "carma/util/ExceptionUtils.h"
#include "carma/util/Trace.h"
#include "carma/util/Logger.h"
#include "carma/util/demangle.h"
#include "carma/util/loggingUtils.h"
#include "carma/util/Backtrace.h"
#include "carma/util/PthreadMutex.h"
#include "carma/util/programLogging.h"
#include "carma/util/Orb.h"
#include "carma/util/ScopedLock.h"

#include <unistd.h>

using namespace ::std;
using namespace carma::util;

using ::log4cpp::Priority;
using ::log4cpp::Category;

namespace {
  ::pthread_mutex_t gOrbRunGuard = PTHREAD_MUTEX_INITIALIZER;
}


// ----------------------------------------------------------------------
// Based on the PushConsumer_impl implementation
// ----------------------------------------------------------------------


NotificationConsumer::NotificationConsumer(Orb* localOrb,
					   string channelName,
					   string consumerName) :
  orb_(localOrb->duplicateOrb()),
  poa_(localOrb->getPOA(channelName+"."+consumerName+"."+"PushConsumer")),
  channelName_ (channelName),
  consumerName_ (consumerName),
  isActive_(false)
{
  localOrb_    = localOrb;
  setProxyPushSupplier();
  activatePoa();
}

NotificationConsumer::~NotificationConsumer( )
try {
    try {
        Category * logger = getProgramLoggerIfAvailable( );
        
        if ( logger != 0 ) {
            *logger << Priority::WARN
                    << "NotificationConsumer::~NotificationConsumer called (\""
                    << channelName_ << "\", \"" << consumerName_ << "\").";
               
            string btText;
            
            {
                Backtrace bt;
                
                bt.captureNoThrow( );
                
                btText = bt.formattedAsString( "    ", "\n" );
            }
    
            *logger << Priority::WARN << "Backtrace is:";
    
            logMultipleLines( *logger, Priority::WARN, btText );
        }
    } catch ( ... ) {
        // Just stifle any exceptions
    }
    
    deactivateConsumer( );
    PortableServer::ObjectId_var oid = poa_->servant_to_id(this);
    poa_->deactivate_object(oid.in());

    if ( !CORBA::is_nil( proxyPushSupplier_ ) ) {
        proxyPushSupplier_->disconnect_structured_push_supplier( );
    }

    // Let the xxx_var manage the ref counting
    proxyPushSupplier_ =
        CosNotifyChannelAdmin::StructuredProxyPushSupplier::_nil();
} catch ( ... ) {
    try {
        programLogErrorIfPossible(
            "Stifling exception in"
            " NotificationConsumer::~NotificationConsumer - " +
            getStringForCaught() );
    } catch ( ... ) {
        // Just stifle any exceptions
    }

    // Just stifle any exceptions
    return;
}


void
NotificationConsumer::activatePoa() {

  try {

    // connect consumer to same POA each time

    const string objString = string( "pushConsumerObject." ) + consumerName_;

    PortableServer::ObjectId_var objectId =
      PortableServer::string_to_ObjectId( objString.c_str( ) );

    poa_->activate_object_with_id( objectId.in( ), this );
    
    localOrb_->activatePOA(localOrb_->getPersistentPOAName(channelName_ + "." + consumerName_ + "." + "PushConsumer"));
  } catch ( const PortableServer::POAManager::AdapterInactive & ) {
    programLogErrorIfPossible( "NotificationConsumer::run - caught "
        "PortableServer::POAManager::AdapterInactive" ); 
    throw;
  }  catch ( const CORBA::SystemException & cse )  {
      ostringstream oss;

      oss << "NotificationConsumer::run - caught CORBA::SystemException ("
          << demangleTypeName( typeid( cse ) ) << ") - " << cse
          << "; reason = " << cse._info().c_str()
          << "; minor code = " << cse.minor();

      programLogErrorIfPossible( oss.str() );
      throw;
  }
}


// IDL to C++ mapping
void
NotificationConsumer::push_structured_event(
    const CosNotification::StructuredEvent & event )
throw ( CORBA::SystemException, CosEventComm::Disconnected )
try {
  /*
    Trace * trace = getProgramTraceObjectIfAvailable( );

    if ( trace != 0 )
        trace->write( Trace::TRACE1, "Event data received!!" );
  */
} catch ( const CORBA::SystemException & ) {
    throw;
} catch ( const CosEventComm::Disconnected & ) {
    throw;
} catch ( ... ) {
    // BULLSHIT_TWC - 14 Sept 2005 - Figure out what to do here
}

void NotificationConsumer::disconnect_structured_push_consumer()
throw ( CORBA::SystemException )
try {

  // the only reason this method is called is because the cleanup from
  // the notification server is trying to shut down the consumer.
  // however, the cleanup is just for "unclean" exits of the consumer,
  // not if it's still actually up.  so if it tries to call this, we
  // create a new consumer and continue to go.

  ScopedLock< ::pthread_mutex_t > lock( gOrbRunGuard );
  if ( !CORBA::is_nil( proxyPushSupplier_ ) ) {
    proxyPushSupplier_->disconnect_structured_push_supplier( );
  }

  // don't release - let the _var handle management
  //  CORBA::release( proxyPushSupplier_ );
  setProxyPushSupplier();
  CosNotifyComm::StructuredPushConsumer_var spc = _this( );
  proxyPushSupplier_->connect_structured_push_consumer( spc );
  // - we don't want a shutdown here since NC can be run in
  //   multi-threaded process
  //orb_->shutdown(false);
} catch ( const CORBA::SystemException & ) {
    throw;
} catch ( ... ) {
    // BULLSHIT_TWC - 14 Sept 2005 - Figure out what to do here
}


void
NotificationConsumer::offer_change(
    const CosNotification::EventTypeSeq & added,
    const CosNotification::EventTypeSeq & removed )
throw( CORBA::SystemException, CosNotifyComm::InvalidEventType )
try {
  try {
    Category *logger = getProgramLoggerIfAvailable();

    if ( logger != 0) {
      *logger << Priority::WARN
	      << "NotificationConsumer::offer_change called (\""
	      << channelName_ << "\" , \"" << consumerName_ << "\").";

      string btText;
      {
	Backtrace bt;
	bt.captureNoThrow( );
	btText = bt.formattedAsString( "    ", "\n" );
      }
      
      *logger << Priority::WARN << "Backtrace is: ";
      logMultipleLines( *logger, Priority::WARN, btText );
    }
  } catch (...) {
    // stifle logging exceptions
  }
  // Intentionally left blank
} catch ( const CORBA::SystemException & ) {
    throw;
} catch ( const CosNotifyComm::InvalidEventType & ) {
    throw;
} catch ( ... ) {
    // BULLSHIT_TWC - 14 Sept 2005 - Figure out what to do here
}


PortableServer::POA_ptr
NotificationConsumer::_default_POA( ) {
    return PortableServer::POA::_duplicate( poa_ );
}


/**.......................................................................
 * A method for calling run if eveng the ProxyPushSupplier object -
 * represents the instance of the notification queue for this consumer
 */
void
NotificationConsumer::setProxyPushSupplier( ) 
{
    proxyPushSupplier_ = localOrb_->getNotifyProxyPushSupplier(channelName_, consumerName_);
}


/**.......................................................................
 * A method to stop the consumer DO
 */
void
NotificationConsumer::deactivateConsumer( ) {
    isActive_ = false;
}


/**.......................................................................
 * A method to test status of the consumer DO
 */
bool
NotificationConsumer::isActive( ) {
    return isActive_ ;
}


/**.......................................................................
 * A method for calling run if event channel is already defined
 */
void
NotificationConsumer::run( )
try {
    Category *logger = getProgramLoggerIfAvailable();
  
    // force servant destruction when out of scope
    // BULLSHIT_TWC - 19 Sept 2005 Figure out what it up with this
    // BULLSHIT_TWC - 19 Sept 2005 Should this at least use _this( )?
    // PortableServer::ServantBase_var tempConsumer = this;

    try {
        CosNotifyComm::StructuredPushConsumer_var spc = _this( );

        proxyPushSupplier_->connect_structured_push_consumer( spc );

        // Now serve the objects (normally blocks here forever).
        isActive_ = true;

        while ( isActive( ) ) {
            // Retrieve and activate the ORB
            if ( CORBA::is_nil( orb_ ) )
              throw CARMA_ERROR("ORB pointer is nil!");

            try {
		// EML: Putting global mutex locks around the global
		// orb already forces serial execution of ANY
		// multi-threaded program using CORBA, and we probably
		// don't want that either, but I certainly don't want
		// it around our private orb

		if(localOrb_->workPending())
                  localOrb_->performWork();

		// EML: For now, I'm putting a 10ms sleep to reduce
		// the load, since we don't expect data on any
		// channel any more frequently than once per
		// half-sec.  We should figure out if there's a
		// better general way to handle the limitations of
		// the CORBA implementation

        struct timespec ts;
                ts.tv_sec  = 0;
                ts.tv_nsec = 10000000;
        nanosleep(&ts, 0);

            } catch (CORBA::BAD_INV_ORDER &) {
                deactivateConsumer();
                programLogWarnIfPossible( "NotificationConsumer::run() - "
                        "Received BAD_INV_ORDER, most likely"
                        " on termination. Deactivating." );
            }

            // 1s sleep exists in work_pending call of OB-4.1.2, but not
            // OB-4.2.0 ... adding another 1ms to this just in case
            // using unix sleep, and upping to 10ms (since that's what
            // it used to be with the old linux scheduler)
            usleep( 10 );

        }

	try {
	  if ( logger != 0) {
	    *logger << Priority::WARN
		    << "NotificationConsumer::run() returning (\""
		    << channelName_ << "\" , \"" << consumerName_ << "\").";
	    
	    string btText;
	    {
	      Backtrace bt;
	      bt.captureNoThrow( );
	      btText = bt.formattedAsString( "    ", "\n" );
	    }
	    
	    *logger << Priority::WARN << "Backtrace is: ";
	    logMultipleLines( *logger, Priority::WARN, btText );
	  }
	} catch (...) {
	  // stifle logging exception
	}
    }  catch ( const CORBA::SystemException & cse )  {
      if ( logger != 0 ) {
        ostringstream oss;

        oss << "NotificationConsumer::run - caught CORBA::SystemException ("
            << demangleTypeName( typeid( cse ) ) << ") - " << cse
            << "; reason = " << cse._info().c_str()
            << "; minor code = " << cse.minor();

        const string msg = oss.str( );

        *logger << Priority::ERROR << msg;
      }

      throw;
    }
} catch ( ... ) {
    try {
        string msg = "NotificationConsumer::run - exception escaping - ";

        msg += getStringForCaught( );

        getProgramLogger( ) << Priority::ERROR << msg;
    } catch ( ... ) {

    }

    throw;
}

File: carma/util/NotificationConsumer.java
// $Id: NotificationConsumer.java,v 1.10 2013/01/31 21:59:02 iws Exp $
// vim: set ts=4 sts=4 sw=4 et:

package carma.util;

import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.PortableServer.POAPackage.*;
import org.omg.CosNotification.*;
import org.omg.CosNotifyChannelAdmin.*;
import org.omg.CosNotifyComm.*;
import java.util.*;
import carma.util.CorbaUtils;

/**
 * A base class that accepts named events from a Push supplier.
 * This is the basis for event driven clients, with the push
 * method being executed when the event is received.  This class
 * is usually extended and the push method is overridden.
 * At the time of construction, this consumer object is bound
 * to a notification channel, and an instance of a notification queue.
 */
public class NotificationConsumer
    extends StructuredPushConsumerPOA implements Runnable {
    protected org.omg.CORBA.ORB orb_;
    protected org.omg.PortableServer.POA poa_;
    protected org.omg.CosNotifyChannelAdmin.StructuredProxyPushSupplier proxyPushSupplier_;


    /**
     * use if multiple consumers in a thread and want to specify own orb
     * @param orb CORBA orb
     * @param channelName name of the event channel
     * @throws org.omg.PortableServer.POAPackage.AdapterAlreadyExists
     * @throws org.omg.PortableServer.POAPackage.InvalidPolicy
     */
    public NotificationConsumer(org.omg.CORBA.ORB orb, String channelName)
        throws org.omg.PortableServer.POAPackage.AdapterAlreadyExists,
               org.omg.PortableServer.POAPackage.InvalidPolicy,
               org.omg.CORBA.ORBPackage.InvalidName {

        // random proxy name
        Random r = new Random();
        String proxyName = Long.toString(Math.abs(r.nextLong()), 36);
        String poaName = channelName + "." + proxyName + "." + "JPushConsumer";

        CorbaUtils cu = new CorbaUtils(orb);

        orb_ = orb;
        poa_ = cu.getPOA(poaName);
        proxyPushSupplier_ = cu.getNotifyProxyPushSupplier(channelName);
    }

    /**
     * Return POA object.  Must be defined for classes inheriting from
     * StructuredPushConsumerPOA
     */
    public org.omg.PortableServer.POA _default_POA() {
        return poa_;
    }

    /**
     * Callback used by ORBacus Notification Server to push events to
     * consumer.  Should be overridden by derived class
     * @throws org.omg.CosEventComm.Disconnected
     */
    public void push_structured_event(org.omg.CosNotification.StructuredEvent event)
        throws org.omg.CosEventComm.Disconnected {

        // if the server receives a Disconnected exception, it will
        // begin a cleanup process on the server-side
        Debug.print(this, Debug.INFO, "NotificationConsumer::push_structured_event: default implementation");
        throw new org.omg.CosEventComm.Disconnected();
    }

    /**
     * Callback method used by the ORBacus Notification.  Should NOT
     * be overridden in derived class
     */
    public void disconnect_structured_push_consumer() {
        byte[] oid = null;

        try {
            oid = poa_.servant_to_id(this);
        } catch (org.omg.PortableServer.POAPackage.ServantNotActive e) {
            e.printStackTrace();
        } catch (org.omg.PortableServer.POAPackage.WrongPolicy e) {
            e.printStackTrace();
        }

        try {
            poa_.deactivate_object(oid);
        } catch (org.omg.PortableServer.POAPackage.ObjectNotActive e) {
            e.printStackTrace();
        } catch (org.omg.PortableServer.POAPackage.WrongPolicy e) {
            e.printStackTrace();
        }

        // stop the current ORB immediately (do not block)
        orb_.shutdown(false);
    }

    /**
     * Method used by the ORBacus notification server to inform the
     * consumer that it will be changing the type of event it will be
     * sending.
     */
    public void offer_change(org.omg.CosNotification.EventType[] added,
                             org.omg.CosNotification.EventType[] removed)
    {
        // intentionally left blank
        Debug.print(this, Debug.INFO, "NotificationConsumer::offer_change called");
    }

    /*------------------------------------------------------------------------*/
    /* Runnable Interface                                                     */
    /*------------------------------------------------------------------------*/

    /**
     * Blocking call that activates the callback for collecting events
     * from the ORBacus notification server.
     */
    public void run() {
        org.omg.PortableServer.POAManager manager = poa_.the_POAManager();

        String objString = "pushConsumerObject";
        byte[] objectId = objString.getBytes();

        try {
            poa_.activate_object_with_id(objectId, this);
        } catch (org.omg.PortableServer.POAPackage.ServantAlreadyActive ex) {
            ex.printStackTrace();
        } catch (org.omg.PortableServer.POAPackage.ObjectAlreadyActive ex) {
            ex.printStackTrace();
        } catch (org.omg.PortableServer.POAPackage.WrongPolicy ex) {
            ex.printStackTrace();
            // do nothing
        }

        org.omg.CosNotifyComm.StructuredPushConsumer structuredPushConsumer = this._this();

        try {
            proxyPushSupplier_.connect_structured_push_consumer(structuredPushConsumer);
        } catch (org.omg.CosEventChannelAdmin.AlreadyConnected ex) {
            ex.printStackTrace();
        } catch (org.omg.CosEventChannelAdmin.TypeError ex) {
            // do nothing
            ex.printStackTrace();
        } catch (org.omg.CORBA.OBJECT_NOT_EXIST ex) {
            ex.printStackTrace();
        }

        try {
            manager.activate();

            // DO is active: block forever on the ORB
            orb_.run();
        } catch (org.omg.PortableServer.POAManagerPackage.AdapterInactive ex) {
            // do nothing
            ex.printStackTrace();
        }
    }

} // end class NotificationConsumer
